moveit_planners/trajopt/include/problem_description.h

```diff
@@ -0,0 +1,286 @@
+  int supported_term_types_;
+};
+
+
+struct ProblemInfo
+{
+public:
+  BasicInfo basic_info;
+  sco::BasicTrustRegionSQPParameters opt_info;
+  std::vector<TermInfoPtr> cost_infos;
+  std::vector<TermInfoPtr> cnt_infos;
+  InitInfo init_info;
+
+  //  tesseract::BasicEnvConstPtr env;
+  //  tesseract::BasicKinConstPtr kin;
+  planning_scene::PlanningSceneConstPtr planning_scene;
+  std::string planning_group_name;
+
+  ProblemInfo(planning_scene::PlanningSceneConstPtr ps, std::string pg)
+    : planning_scene(ps), planning_group_name(pg) {}
+  //  void fromJson(const Json::Value& v);
+
+private:
+  // void readBasicInfo(const Json::Value& v);
+  // void readOptInfo(const Json::Value& v);
+  // void readCosts(const Json::Value& v);
+  // void readConstraints(const Json::Value& v);
+  //  void readInitInfo(const Json::Value& v);
+};
+
+
+/**
+ * Holds all the data for a trajectory optimization problem
+ * so you can modify it programmatically, e.g. add your own costs
+ */
+class  TrajOptProblem :  public sco::OptProb
+{
+public:
+  TrajOptProblem();
+  TrajOptProblem(const ProblemInfo& problem_info);
+  virtual ~TrajOptProblem() = default;
+  sco::VarVector GetVarRow(int i, int start_col, int num_col) { return m_traj_vars.rblock(i, start_col, num_col); }
+  sco::VarVector GetVarRow(int i) { return m_traj_vars.row(i); }
+  sco::Var& GetVar(int i, int j) { return m_traj_vars.at(i, j); }
+  trajopt::VarArray& GetVars() { return m_traj_vars; }
+  /** @brief Returns the number of steps in the problem. This is the number of rows in the optimization matrix.*/
+  int GetNumSteps() { return m_traj_vars.rows(); }
+  /** @brief Returns the problem DOF. This is the number of columns in the optization matrix.
+   * Note that this is not necessarily the same as the kinematic DOF.*/
+  int GetNumDOF() { return m_traj_vars.cols(); }
+  // tesseract::BasicKinConstPtr GetKin() { return m_kin; }
+  //tesseract::BasicEnvConstPtr GetEnv() { return m_env; }
+  planning_scene::PlanningSceneConstPtr GetPlanningScene() { return planning_scene_;}
+  void SetInitTraj(const trajopt::TrajArray& x){  m_init_traj = x; }
+  trajopt::TrajArray GetInitTraj() { return m_init_traj; }
+  //  friend TrajOptProbPtr ConstructProblem(const ProblemConstructionInfo&);
+  /** @brief Returns TrajOptProb.has_time */
+  bool GetHasTime() { return has_time; }
+  /** @brief Sets TrajOptProb.has_time  */
+  void SetHasTime(bool tmp) { has_time = tmp; }
+
+private:
+  /** @brief If true, the last column in the optimization matrix will be 1/dt */
+  bool has_time;
+  trajopt::VarArray m_traj_vars;
+  // tesseract::BasicKinConstPtr m_kin;
+  // tesseract::BasicEnvConstPtr m_env;
+  planning_scene::PlanningSceneConstPtr planning_scene_;
+  std::string planning_group_;
+  trajopt::TrajArray m_init_traj;
+};
+
+
+/** @brief This term is used when the goal frame is fixed in cartesian space
+
+  Set term_type == TT_COST or TT_CNT for cost or constraint.
+*/
+struct CartPoseTermInfo : public TermInfo
```

Comment:
Do you need this struct current? Since you arent' supporting cartesian planning yet?

Reply:
Actually the cartesian constrain is implemented too but I am not sure about its validity yet. I make sure it is working correctly, otherwise will remove it and make another PR for that.